src/filesystem_helper.cpp

```diff
@@ -0,0 +1,234 @@
+#  include <io.h>
+#  define access _access_s
+#else
+#  include <dirent.h>
+#  include <sys/types.h>
+#  include <unistd.h>
+#endif
+
+#include "rcutils/get_env.h"
+
+namespace rcpputils
+{
+namespace fs
+{
+
+bool path::exists() const
+{
+  return access(path_.c_str(), 0) == 0;
+}
+
+bool path::is_directory() const noexcept
+{
+  struct stat stat_buffer;
+  const auto rc = stat(path_.c_str(), &stat_buffer);
+
+  if (rc != 0) {
+    return false;
+  }
+
+#ifdef _WIN32
+  return (stat_buffer.st_mode & S_IFDIR) == S_IFDIR;
+#else
+  return S_ISDIR(stat_buffer.st_mode);
+#endif
+}
+
+bool path::is_regular_file() const noexcept
+{
+  struct stat stat_buffer;
+  const auto rc = stat(path_.c_str(), &stat_buffer);
+
+  if (rc != 0) {
+    return false;
+  }
+
+#ifdef _WIN32
+  return (stat_buffer.st_mode & S_IFREG) == S_IFREG;
+#else
+  return S_ISREG(stat_buffer.st_mode);
+#endif
+}
+
+path temp_directory_path()
+{
+#ifdef _WIN32
+#ifdef UNICODE
+#error "rcpputils::fs does not support Unicode paths"
+#endif
+  TCHAR temp_path[MAX_PATH];
+  DWORD size = GetTempPathA(MAX_PATH, temp_path);
+  if (size > MAX_PATH || size == 0) {
+    std::error_code ec(static_cast<int>(GetLastError()), std::system_category());
+    throw std::system_error(ec, "cannot get temporary directory path");
+  }
+  temp_path[size] = '\0';
+#else
+  const char * temp_path = NULL;
+  const char * ret_str = rcutils_get_env("TMPDIR", &temp_path);
+  if (NULL != ret_str || *temp_path == '\0') {
+    temp_path = "/tmp";
+  }
+#endif
+  return path(temp_path);
+}
+
+path current_path()
+{
+#ifdef _WIN32
+#ifdef UNICODE
+#error "rcpputils::fs does not support Unicode paths"
+#endif
+  char cwd[MAX_PATH];
+  if (nullptr == _getcwd(cwd, MAX_PATH)) {
+#else
+  char cwd[PATH_MAX];
+  if (nullptr == getcwd(cwd, PATH_MAX)) {
+#endif
+    std::error_code ec{errno, std::system_category()};
+    errno = 0;
+    throw std::system_error{ec, "cannot get current working directory"};
+  }
+
+  return path(cwd);
+}
+
+bool create_directories(const path & p)
+{
+  path p_built;
+  int status = 0;
+
+  for (auto it = p.cbegin(); it != p.cend() && status == 0; ++it) {
+    if (!p_built.empty() || it->empty()) {
+      p_built /= *it;
+    } else {
+      p_built = *it;
+    }
+    if (!p_built.exists()) {
+#ifdef _WIN32
+      status = _mkdir(p_built.string().c_str());
+#else
+      status = mkdir(p_built.string().c_str(), S_IRWXU | S_IRWXG | S_IRWXO);
+#endif
+    }
+  }
+  return status == 0 && p_built.is_directory();
+}
+
+bool remove(const path & p)
+{
+#ifdef _WIN32
+  struct _stat s;
+  if (_stat(p.string().c_str(), &s) == 0) {
+    if (s.st_mode & S_IFDIR) {
+      return _rmdir(p.string().c_str()) == 0;
+    }
+    if (s.st_mode & S_IFREG) {
+      return ::remove(p.string().c_str()) == 0;
+    }
+  }
+  return false;
+#else
+  return ::remove(p.string().c_str()) == 0;
+#endif
+}
+
+bool remove_all(const path & p)
+{
+  if (!is_directory(p)) {return rcpputils::fs::remove(p);}
```

Comment:
Nit: I know you are just moving this code around, but can we make this more like our usual style?

```suggestion
  if (!is_directory(p)) {
    return rcpputils::fs::remove(p);
  }
```

Reply:
done in https://github.com/ros2/rcpputils/pull/118/commits/262bef86ab565de506a683d9f71b3e45cdc24789